src/Bedrock.Framework.Experimental/Protocols/Memcached/MemcachedResponse.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Text;
+using static Bedrock.Framework.Experimental.Protocols.Memcached.Enums;
+
+namespace Bedrock.Framework.Experimental.Protocols.Memcached
+{
+    public class MemcachedResponse
+    {
+        public ReadOnlyMemory<byte> Data { get; private set; }
+        public TypeCode Flags { get; set; }
+        public MemcachedResponseHeader Header { get; private set; }
+        public void ReadHeader(ReadOnlySpan<byte> buffer)
+        {
+           
+            if (buffer[0] != MemcachedResponseHeader.Magic)
+                throw new ArgumentException("Magic mismatch");
+
+            this.Header = new MemcachedResponseHeader()
+            {
+                Opcode = (Opcode)buffer[1],
+                KeyLength = BinaryPrimitives.ReadUInt16BigEndian(buffer.Slice(2)),
+                ExtraLength = buffer[4],
+                DataType = buffer[5],
+                ResponseStatus = (ResponseStatus)BinaryPrimitives.ReadUInt16BigEndian(buffer.Slice(6)),
+                TotalBodyLength = BinaryPrimitives.ReadUInt32BigEndian(buffer.Slice(8)),
+                Opaque = BinaryPrimitives.ReadUInt32BigEndian(buffer.Slice(12)),
+                Cas = BinaryPrimitives.ReadUInt64BigEndian(buffer.Slice(16)),
+            };
+        }
+        public void ReadBody(ReadOnlySequence<byte> sequence)
+        {
+            if (sequence.Length == 0)
+                return;
+            if (sequence.IsSingleSegment)
+            {
+                Flags = (TypeCode)BinaryPrimitives.ReadUInt32BigEndian(sequence.First.Span);
+                Data = sequence.First.Slice(Header.KeyLength + Header.ExtraLength);
+            }
+            else
+            {
+                var buffer = MemoryPool<byte>.Shared.Rent((int)Header.TotalBodyLength).Memory.Slice(0, (int)Header.TotalBodyLength);
```

Comment:
Don't copy here, just slice the existing sequence. It's the caller's job to call Advance when they are done with the memory. See the other message readers as an example.

Reply:
Fixed [https://github.com/davidfowl/BedrockFramework/pull/84/commits/8e7d6f6ac61db05f35b4fcf847d5cb95444dc964](https://github.com/davidfowl/BedrockFramework/pull/84/commits/8e7d6f6ac61db05f35b4fcf847d5cb95444dc964)